src/snowflake/snowpark/mock/_stage_registry.py

```diff
@@ -0,0 +1,232 @@
+
+            file_name = os.path.basename(local_file_name)
+            stage_target_dir_path = os.path.join(self._working_directory, stage_prefix)
+
+            if not os.path.exists(stage_target_dir_path):
+                os.makedirs(stage_target_dir_path)
+
+            if (
+                os.path.isfile(os.path.join(stage_target_dir_path, file_name))
+                and not overwrite
+            ):
+                status = "SKIPPED"
+            else:
+                shutil.copy(local_file_name, stage_target_dir_path)
+                status = "UPLOADED"
+
+            file_size = os.path.getsize(os.path.expanduser(local_file_name))
+            result_df.loc[len(result_df)] = {
+                k: v
+                for k, v in zip(
+                    PUT_RESULT_KEYS,
+                    [
+                        file_name,
+                        file_name,
+                        file_size,
+                        file_size,
+                        "NONE",
+                        "NONE",
+                        status,
+                        "",
+                    ],
+                )
+            }
+        return result_df
+
+    def upload_stream(
+        self,
+        input_stream: IO[bytes],
+        stage_prefix: str,
+        file_name: str,
+        overwrite: bool = False,
+    ) -> Dict:
+        stage_target_dir_path = os.path.join(self._working_directory, stage_prefix)
+
+        if not os.path.exists(stage_target_dir_path):
+            os.makedirs(stage_target_dir_path)
+
+        status = "UPLOADED"
+        if (
+            os.path.isfile(os.path.join(stage_target_dir_path, file_name))
+            and not overwrite
+        ):
+            status = "SKIPPED"
+        else:
+            with open(os.path.join(stage_target_dir_path, file_name), "wb") as f:
+                try:
+                    f.write(input_stream.read())
+                except ValueError as exc:
+                    raise SnowparkUploadFileException(
+                        message="[Local Testing] Reading closed file while uploading stream",
+                        error_code="1408",
+                    ) from exc
+
+        file_size = os.path.getsize(os.path.join(stage_target_dir_path, file_name))
+        return {
+            "data": [
+                (file_name, file_name, file_size, file_size, "NONE", "NONE", status, "")
+            ],
+            "sfqid": None,
+        }
+
+
+class StageEntityRegistry:
+    # Registry to store tables and views.
+    def __init__(self, conn: "MockServerConnection") -> None:
+        self._root_dir = tempfile.TemporaryDirectory()
+        self._stage_registry = {}
+
+    def create_or_replace_stage(self, stage_name):
+        self._stage_registry[stage_name] = StageEntity(self._root_dir.name, stage_name)
+
+    def __getitem__(self, stage_name: str):
+        # the assumption here is that stage always exists
+        return self._stage_registry[stage_name]
```

Comment:
Users lack the API to create stages, if we assume the stage always exists, we need to create the corresponding StageEntity here if it doesn't exist in the map.

Reply:
good catch, updated